src/Common/FileCache.cpp

```diff
@@ -598,6 +635,163 @@ bool LRUFileCache::tryReserve(
     return true;
 }
 
+LRUFileCache::ReserveResult LRUFileCache::tryReserveForQuery(const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock)
+{
+    if (!CurrentThread::isInitialized() || !CurrentThread::get().getQueryContext() || CurrentThread::getQueryId().size == 0)
+        return ReserveResult::NO_NEED;
+
+    auto query_id = CurrentThread::getQueryId().toString();
+    auto query_iter = query_map.find(query_id);
+
+    if (query_iter == query_map.end())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cache became inconsistent. There must be a bug");
+
+    auto query_context = query_iter->second;
+
+    /// The maximum cache capacity of the request is not reached, thus the
+    //// cache block is evicted from the main LRU queue by tryReserveForMainList().
+    if (query_context->cache_size + size < query_context->max_cache_size)
+    {
+        return ReserveResult::NO_NEED;
+    }
+    /// The maximum cache size of the query is reached, the cache will be 
+    /// evicted from the history cache accessed by the current query.
```

Comment:
My implementation is to evict only the cache written by the query, and the queue in the context only records the query as the downloader's cache segment information.

Reply:
The context of the query is updated only when tryReserve() is called, and the premise of calling tryReserve() is that the query is the downloader of the file_segment

Reply:
> My implementation is to evict only the cache written by the query, and the queue in the context only records the query as the downloader's cache segment information.

Yes, I understood, what I meant is -- if we reached the limit of cache for the query, then may be just stop caching without evicting cache written by current query.

Reply:
I don't know if my understanding is correct, I think you mean that when the maximum cache size for a query is reached, we only stop the cache writes for that query. 
This approach is indeed simpler and even we don't need to maintain every query. However, this approach may lose the hot spot information within a single query. For example, for a query, I want to keep the frequently accessed cache as much as possible. Do you think this makes sense? If not maybe we can take your suggestion.

Reply:
Yes, this is what I meant. I do agree that your approach is good and makes sense to have, so may be add a setting to control this behaviour so both options are supported? It is very easy to add, but might be beneficial.

Reply:
Ok, no problem, I plan to set a variable **skip_download_if_exceeds_query_cache**, when the variable is 1, we read the data directly from the remote fs instead of downloading to the local, this variable is independent for each query.